src/json.hpp

```diff
@@ -1225,6 +1445,21 @@ class basic_json
         assert_invariant();
     }
 
+    // constructor chosen when JSONSerializer::to_json exists for type T
+    template <typename T, enable_if_t<detail::has_to_json<
+                              JSONSerializer, basic_json, uncvref_t<T>>::value and not (detail::is_compatible_object_type<object_t, uncvref_t<T>>::value or
+                                                                                        detail::is_compatible_array_type<basic_json_t, uncvref_t<T>>::value or
+                                                                                        detail::is_compatible_float_type<number_float_t, uncvref_t<T>>::value or
+                                                                                        detail::is_compatible_integer_type<number_integer_t, uncvref_t<T>>::value or
+                                                                                        detail::is_compatible_unsigned_integer_type<number_unsigned_t, uncvref_t<T>>::value or
+                                                                                        std::is_constructible<string_t, uncvref_t<T>>::value or
+                                                                                        std::is_base_of<std::istream, uncvref_t<T>>::value or
+                                                                                        std::is_same<boolean_t, uncvref_t<T>>::value), int> = 0>
```

Comment:
Might want to abstract all these checks into an `detail::is_basic_json_compatible<uncvref_t<T>>` or something like that. It might make the function definition clearer, and it also might make it easier to explain what is going on.